include/hydrogen/device/gpu/cuda/SyncInfo.hpp

```diff
@@ -0,0 +1,89 @@
+#ifndef HYDROGEN_DEVICE_GPU_CUDA_SYNCINFO_HPP_
+#define HYDROGEN_DEVICE_GPU_CUDA_SYNCINFO_HPP_
+
+#include <cuda_runtime_api.h>
+
+#include <hydrogen/SyncInfo.hpp>
+#include <hydrogen/meta/MetaUtilities.hpp>
+
+#include "CUDAError.hpp"
+#include "CUDAManagement.hpp"
+
+namespace hydrogen
+{
+
+template <>
+class SyncInfo<Device::GPU>
+{
+public:
+    SyncInfo()
+        : SyncInfo{cuda::GetDefaultStream(), cuda::GetDefaultEvent()}
+    {}
+
+    SyncInfo(cudaStream_t stream, cudaEvent_t event)
+        : stream_{stream}, event_{event}
+    {}
+
+    void Merge(SyncInfo<Device::GPU> const& si) noexcept
+    {
+        if (si.stream_)
+            stream_ = si.stream_;
+        if (si.event_)
+            event_ = si.event_;
+    }
+
+    cudaStream_t Stream() const noexcept { return stream_; }
+    cudaEvent_t Event() const noexcept { return event_; }
+private:
+    friend void DestroySyncInfo(SyncInfo<Device::GPU>&);
+    cudaStream_t stream_;
+    cudaEvent_t event_;
+};// struct SyncInfo<Device::GPU>
+
+inline void AddSynchronizationPoint(SyncInfo<Device::GPU> const& syncInfo)
+{
+    H_CHECK_CUDA(cudaEventRecord(syncInfo.Event(), syncInfo.Stream()));
+}
+
+
+namespace details
+{
+inline void AddSyncPoint(
+    SyncInfo<Device::CPU> const& master,
+    SyncInfo<Device::GPU> const& dependent)
+{
+    // The GPU must wait for the CPU.
+    // This probably requires Aluminum, but it might also just be a logic error.
+    throw std::logic_error(
+        "Either this function doesn't make sense "
+        "or it should use Al::GPUWait.");
```

Comment:
Yes, trivial synchronization is the right behavior for a single-threaded application. Since we don't support multiple threads in Hydrogen, this is fine.